File: game/Effects.ts
// 效果模型：一次连线的结算产物是效果列表 [{kind, value, target}]。
// 注意：持续回合等时间性质不进入三元组，统一放在 Config 的持续时间表里。

import { Config } from 'game/Config';

/** 执行效果的种类。新增效果种类 = 注册一条 handler（本直文件只列枚举）。 */
export const enum EffectKind {
	PhysicalDamage = 'physicalDamage',
	MagicDamage = 'magicDamage',
	Heal = 'heal',
	Dispel = 'dispel',
	Shield = 'shield',
	BuffDamage = 'buffDamage',
	DebuffArmor = 'debuffArmor',
	ManaGain = 'manaGain',
	BoardBlock = 'boardBlock',
	BoardShuffle = 'boardShuffle',
	BoardBlast = 'boardBlast',
}

/** 效果作用对象。`Reserved` 为预留未实现的枚举位。 */
export const enum EffectTarget {
	Self = 'self',
	CurrentEnemy = 'currentEnemy',
	AllEnemies = 'allEnemies',
	Board = 'board',
	Reserved = 'reserved',
}

/** 严格三元组：执行效果 / 执行数值 / 执行对象。不带可选字段，A避免 Lua 数组存 undefined。 */
export interface EffectSpec {
	kind: EffectKind;
	value: number;
	target: EffectTarget;
}

/** 纯数据规则：“链长 → 数值”也是数据，不是函数。 */
export interface EffectRule {
	kind: EffectKind;
	target: EffectTarget;
	/** 链长下限时的数值。 */
	base: number;
	/** 每多一块的增量（可为 0）。 */
	perBlock: number;
	/** 生效所需最小链长。 */
	minChain: number;
	/** 取整粒度：0 = 四舍五入；>0 = 向下取整到该粒度。 */
	floorTo: number;
}

/** 构造一条效果规则，避免各处重复写全字段。 */
export function makeRule(kind: EffectKind, target: EffectTarget, base: number, perBlock: number, minChain: number = Config.MinChainLength, floorTo: number = 0): EffectRule {
	return { kind, target, base, perBlock, minChain, floorTo };
}

/** 按取整粒度把原始数值归整。 */
export function roundRuleValue(raw: number, floorTo: number): number {
	if (floorTo > 0) {
		return Math.floor(raw / floorTo) * floorTo;
	}
	return Math.round(raw);
}

/**
 * 把方块的效果规则展开为效果三元组列表。
 * 链长低于规则下限时跳过该条；链长低于全局下限时统一返回空列表。
 */
export function resolveEffects(rules: EffectRule[], chainLength: number): EffectSpec[] {
	const specs: EffectSpec[] = [];
	if (chainLength < Config.MinChainLength) {
		return specs;
	}
	for (const rule of rules) {
		if (chainLength < rule.minChain) {
			continue;
		}
		const raw = rule.base + rule.perBlock * chainLength;
		specs.push({ kind: rule.kind, value: roundRuleValue(raw, rule.floorTo), target: rule.target });
	}
	return specs;
}

/** 效果类型的中文名（用于日志、飘字与自检报告）。 */
export const EffectKindNames: Record<EffectKind, string> = {
	[EffectKind.PhysicalDamage]: '物理伤害',
	[EffectKind.MagicDamage]: '魔法伤害',
	[EffectKind.Heal]: '治疗',
	[EffectKind.Dispel]: '净化',
	[EffectKind.Shield]: '护盾',
	[EffectKind.BuffDamage]: '增伤',
	[EffectKind.DebuffArmor]: '破甲',
	[EffectKind.ManaGain]: '魔力',
	[EffectKind.BoardBlock]: '封锁',
	[EffectKind.BoardShuffle]: '洗牌',
	[EffectKind.BoardBlast]: '引爆',
};

/** 作用对象的中文名。 */
export const EffectTargetNames: Record<EffectTarget, string> = {
	[EffectTarget.Self]: '自身',
	[EffectTarget.CurrentEnemy]: '当前敌人',
	[EffectTarget.AllEnemies]: '全体敌人',
	[EffectTarget.Board]: '棋盘',
	[EffectTarget.Reserved]: '预留',
};

/** 把效果列表格式化为可读文本：[物理伤害 13 → 当前敌人] */
export function formatEffects(specs: EffectSpec[]): string {
	const parts: string[] = [];
	for (const spec of specs) {
		parts.push(EffectKindNames[spec.kind] + ' ' + spec.value + ' → ' + EffectTargetNames[spec.target]);
	}
	return parts.join(' / ');
}

/** 战斗单位状态（玩家与敌人共用）。 */
export interface ActorState {
	hp: number;
	maxHp: number;
	shield: number;
	armor: number;
	buffStacks: number;
	debuffStacks: number;
	mana: number;
	maxMana: number;
}

/** 执行器需要的棋盘能力；由 Board 结构化实现，避免模块循环依赖。 */
export interface BoardOps {
	/** 重排整盘。 */
	shuffle(): void;
	/** 清除场上最大的连通块，返回移除块数。 */
	blastLargestGroup(): number;
	/** 把 count 个格子封锁为固定障碍（不可消除、不可连线的封锁格），返回实际封锁数。 */
	blockCells(count: number): number;
}

/** 执行一次效果列表所需的上下文。 */
export interface EffectContext {
	player: ActorState;
	currentEnemy: ActorState;
	allEnemies: ActorState[];
	board: BoardOps;
}

/** 新建一个战斗单位状态。 */
export function makeActor(hp: number, armor: number, maxMana: number): ActorState {
	return { hp, maxHp: hp, shield: 0, armor, buffStacks: 0, debuffStacks: 0, mana: 0, maxMana };
}

/** 叠加魔力并夹在 [0, maxMana] 内，返回实际增加量。 */
export function addMana(actor: ActorState, amount: number): number {
	const before = actor.mana;
	actor.mana += Math.round(amount);
	if (actor.mana > actor.maxMana) {
		actor.mana = actor.maxMana;
	}
	if (actor.mana < 0) {
		actor.mana = 0;
	}
	return actor.mana - before;
}

/** 扣除魔力；魔力不足时返回 false 且不扣除。 */
export function spendMana(actor: ActorState, amount: number): boolean {
	if (actor.mana < amount) {
		return false;
	}
	actor.mana -= amount;
	return true;
}

/** 破甲减益后的有效护甲。 */
export function effectiveArmor(actor: ActorState): number {
	const armor = actor.armor * (1 - actor.debuffStacks * Config.DebuffArmorPerStack);
	return armor < 0 ? 0 : armor;
}

/** 增伤层数换算的伤害倍率。 */
export function damageMultiplier(actor: ActorState): number {
	return 1 + actor.buffStacks * Config.BuffDamagePerStack;
}

/** 先扣护盾再扣生命，返回实际造成的生命伤害。 */
export function applyDamage(actor: ActorState, amount: number): number {
	let damage = Math.round(amount);
	if (damage < Config.MinDamage) {
		damage = Config.MinDamage;
	}
	if (actor.shield > 0) {
		const absorbed = damage < actor.shield ? damage : actor.shield;
		actor.shield -= absorbed;
		damage -= absorbed;
	}
	if (damage > 0) {
		actor.hp -= damage;
		if (actor.hp < 0) {
			actor.hp = 0;
		}
	}
	return damage;
}

/** 目标解析表：target → 接收者列表。新增目标类型只需在此加一条。 */
const targetResolvers: Record<string, ((context: EffectContext) => ActorState[]) | undefined> = {
	[EffectTarget.Self]: (context) => [context.player],
	[EffectTarget.CurrentEnemy]: (context) => [context.currentEnemy],
	[EffectTarget.AllEnemies]: (context) => context.allEnemies,
};

type EffectHandler = (context: EffectContext, spec: EffectSpec) => void;

/** 按 target 取接收者；未注册或预留的目标返回空列表。 */
export function resolveTargets(context: EffectContext, target: EffectTarget): ActorState[] {
	const resolver = targetResolvers[target];
	if (resolver === undefined) {
		return [];
	}
	return resolver(context);
}

/**
 * 效果执行表：kind → handler。新增效果种类只需在此加一条，
 * 方块效果与技能效果共用本表，执行器核心不含 if (kind === ...) 分支。
 */
const effectHandlers: Record<string, EffectHandler | undefined> = {
	[EffectKind.PhysicalDamage]: (context, spec) => {
		for (const target of resolveTargets(context, spec.target)) {
			applyDamage(target, spec.value * damageMultiplier(context.player) - effectiveArmor(target));
		}
	},
	[EffectKind.MagicDamage]: (context, spec) => {
		for (const target of resolveTargets(context, spec.target)) {
			const bonus = target.shield > 0 ? 1 + Config.MagicShieldBonus : 1;
			applyDamage(target, spec.value * bonus * damageMultiplier(context.player));
		}
	},
	[EffectKind.Heal]: (context, spec) => {
		for (const target of resolveTargets(context, spec.target)) {
			target.hp += Math.round(spec.value);
			if (target.hp > target.maxHp) {
				target.hp = target.maxHp;
			}
		}
	},
	[EffectKind.Dispel]: (context, spec) => {
		for (const target of resolveTargets(context, spec.target)) {
			target.debuffStacks -= Math.round(spec.value);
			if (target.debuffStacks < 0) {
				target.debuffStacks = 0;
			}
		}
	},
	[EffectKind.Shield]: (context, spec) => {
		for (const target of resolveTargets(context, spec.target)) {
			target.shield += Math.round(spec.value);
		}
	},
	[EffectKind.BuffDamage]: (context, spec) => {
		for (const target of resolveTargets(context, spec.target)) {
			target.buffStacks += Math.round(spec.value);
		}
	},
	[EffectKind.DebuffArmor]: (context, spec) => {
		for (const target of resolveTargets(context, spec.target)) {
			target.debuffStacks += Math.round(spec.value);
		}
	},
	[EffectKind.ManaGain]: (context, spec) => {
		for (const target of resolveTargets(context, spec.target)) {
			addMana(target, spec.value);
		}
	},
	[EffectKind.BoardShuffle]: (context, _spec) => {
		context.board.shuffle();
	},
	[EffectKind.BoardBlast]: (context, _spec) => {
		context.board.blastLargestGroup();
	},
	[EffectKind.BoardBlock]: (context, spec) => {
		context.board.blockCells(Math.round(spec.value));
	},
};

/** 尚未注册 handler 的效果次数（供自检与扩展性验收观测）。 */
let unhandledEffects = 0;

export function unhandledEffectCount(): number {
	return unhandledEffects;
}

export function resetUnhandledEffectCount(): void {
	unhandledEffects = 0;
}

/**
 * 统一执行器：只做“查表 → 按 target 取接收者 → 调用 handler”，
 * 返回实际执行的效果条目数；未注册的 kind 计入告警。
 */
export function executeEffects(specs: EffectSpec[], context: EffectContext): number {
	let executed = 0;
	for (const spec of specs) {
		const handler = effectHandlers[spec.kind];
		if (handler === undefined) {
			unhandledEffects++;
			continue;
		}
		handler(context, spec);
		executed++;
	}
	return executed;
}

File: game/Effects.lua
-- [ts]: Effects.ts
local ____exports = {} -- 1
local ____ChainTiers = require("game.ChainTiers") -- 4
local ChainTiers = ____ChainTiers.ChainTiers -- 4
local ____Config = require("game.Config") -- 5
local Config = ____Config.Config -- 5
--- 构造一条效果规则，避免各处重复写全字段。
function ____exports.makeRule(kind, target, base, perBlock, minChain, floorTo, scaled) -- 55
	if minChain == nil then -- 55
		minChain = Config.MinChainLength -- 55
	end -- 55
	if floorTo == nil then -- 55
		floorTo = 0 -- 55
	end -- 55
	if scaled == nil then -- 55
		scaled = true -- 55
	end -- 55
	return { -- 56
		kind = kind, -- 56
		target = target, -- 56
		base = base, -- 56
		perBlock = perBlock, -- 56
		minChain = minChain, -- 56
		floorTo = floorTo, -- 56
		scaled = scaled -- 56
	} -- 56
end -- 55
--- 按取整粒度把原始数值归整。
function ____exports.roundRuleValue(raw, floorTo) -- 60
	if floorTo > 0 then -- 60
		return math.floor(raw / floorTo) * floorTo -- 62
	end -- 62
	return math.floor(raw + 0.5) -- 64
end -- 60
--- 把方块的效果规则展开为效果三元组列表。
-- 链长低于规则下限时跳过该条；链长低于全局下限时统一返回空列表。
function ____exports.resolveEffects(rules, chainLength) -- 71
	local specs = {} -- 72
	if chainLength < Config.MinChainLength then -- 72
		return specs -- 74
	end -- 74
	local tierMultiplier = ChainTiers:multiplierOf(chainLength) -- 77
	for ____, rule in ipairs(rules) do -- 78
		do -- 78
			if chainLength < rule.minChain then -- 78
				goto __continue7 -- 80
			end -- 80
			local raw = rule.base + rule.perBlock * chainLength -- 82
			if rule.scaled then -- 82
				raw = raw * tierMultiplier -- 84
			end -- 84
			specs[#specs + 1] = { -- 86
				kind = rule.kind, -- 86
				value = ____exports.roundRuleValue(raw, rule.floorTo), -- 86
				target = rule.target -- 86
			} -- 86
		end -- 86
		::__continue7:: -- 86
	end -- 86
	return specs -- 88
end -- 71
--- 效果类型的中文名（用于日志、飘字与自检报告）。
____exports.EffectKindNames = { -- 92
	physicalDamage = "物理伤害", -- 93
	magicDamage = "魔法伤害", -- 94
	heal = "治疗", -- 95
	dispel = "净化", -- 96
	shield = "护盾", -- 97
	buffDamage = "增伤", -- 98
	debuffArmor = "破甲", -- 99
	manaGain = "魔力", -- 100
	boardBlock = "封锁", -- 101
	boardShuffle = "洗牌", -- 102
	boardBlast = "引爆" -- 103
} -- 103
--- 作用对象的中文名。
____exports.EffectTargetNames = { -- 107
	self = "自身", -- 108
	currentEnemy = "当前敌人", -- 109
	allEnemies = "全体敌人", -- 110
	board = "棋盘", -- 111
	reserved = "预留" -- 112
} -- 112
--- 把效果列表格式化为可读文本：[物理伤害 13 → 当前敌人]
function ____exports.formatEffects(specs) -- 116
	local parts = {} -- 117
	for ____, spec in ipairs(specs) do -- 118
		parts[#parts + 1] = (((____exports.EffectKindNames[spec.kind] .. " ") .. tostring(spec.value)) .. " → ") .. ____exports.EffectTargetNames[spec.target] -- 119
	end -- 119
	return table.concat(parts, " / ") -- 121
end -- 116
--- 新建一个战斗单位状态。
function ____exports.makeActor(hp, armor, maxMana) -- 155
	return { -- 156
		hp = hp, -- 156
		maxHp = hp, -- 156
		shield = 0, -- 156
		armor = armor, -- 156
		buffStacks = 0, -- 156
		debuffStacks = 0, -- 156
		mana = 0, -- 156
		maxMana = maxMana -- 156
	} -- 156
end -- 155
--- 叠加魔力并夹在 [0, maxMana] 内，返回实际增加量。
function ____exports.addMana(actor, amount) -- 160
	local before = actor.mana -- 161
	actor.mana = actor.mana + math.floor(amount + 0.5) -- 162
	if actor.mana > actor.maxMana then -- 162
		actor.mana = actor.maxMana -- 164
	end -- 164
	if actor.mana < 0 then -- 164
		actor.mana = 0 -- 167
	end -- 167
	return actor.mana - before -- 169
end -- 160
--- 扣除魔力；魔力不足时返回 false 且不扣除。
function ____exports.spendMana(actor, amount) -- 173
	if actor.mana < amount then -- 173
		return false -- 175
	end -- 175
	actor.mana = actor.mana - amount -- 177
	return true -- 178
end -- 173
--- 破甲减益后的有效护甲。
function ____exports.effectiveArmor(actor) -- 182
	local armor = actor.armor * (1 - actor.debuffStacks * Config.DebuffArmorPerStack) -- 183
	return armor < 0 and 0 or armor -- 184
end -- 182
--- 增伤层数换算的伤害倍率。
function ____exports.damageMultiplier(actor) -- 188
	return 1 + actor.buffStacks * Config.BuffDamagePerStack -- 189
end -- 188
--- 先扣护盾再扣生命，返回实际造成的生命伤害。
function ____exports.applyDamage(actor, amount) -- 193
	local damage = math.floor(amount + 0.5) -- 194
	if damage < Config.MinDamage then -- 194
		damage = Config.MinDamage -- 196
	end -- 196
	if actor.shield > 0 then -- 196
		local absorbed = damage < actor.shield and damage or actor.shield -- 199
		actor.shield = actor.shield - absorbed -- 200
		damage = damage - absorbed -- 201
	end -- 201
	if damage > 0 then -- 201
		actor.hp = actor.hp - damage -- 204
		if actor.hp < 0 then -- 204
			actor.hp = 0 -- 206
		end -- 206
	end -- 206
	return damage -- 209
end -- 193
--- 目标解析表：target → 接收者列表。新增目标类型只需在此加一条。
local targetResolvers = { -- 213
	self = function(context) return {context.player} end, -- 214
	currentEnemy = function(context) return {context.currentEnemy} end, -- 215
	allEnemies = function(context) return context.allEnemies end -- 216
} -- 216
--- 按 target 取接收者；未注册或预留的目标返回空列表。
function ____exports.resolveTargets(context, target) -- 222
	local resolver = targetResolvers[target] -- 223
	if resolver == nil then -- 223
		return {} -- 225
	end -- 225
	return resolver(context) -- 227
end -- 222
--- 效果执行表：kind → handler。新增效果种类只需在此加一条，
-- 方块效果与技能效果共用本表，执行器核心不含 if (kind === ...) 分支。
local effectHandlers = { -- 234
	physicalDamage = function(context, spec) -- 235
		for ____, target in ipairs(____exports.resolveTargets(context, spec.target)) do -- 236
			____exports.applyDamage( -- 237
				target, -- 237
				spec.value * ____exports.damageMultiplier(context.player) - ____exports.effectiveArmor(target) -- 237
			) -- 237
		end -- 237
	end, -- 235
	magicDamage = function(context, spec) -- 240
		for ____, target in ipairs(____exports.resolveTargets(context, spec.target)) do -- 241
			local bonus = target.shield > 0 and 1 + Config.MagicShieldBonus or 1 -- 242
			____exports.applyDamage( -- 243
				target, -- 243
				spec.value * bonus * ____exports.damageMultiplier(context.player) -- 243
			) -- 243
		end -- 243
	end, -- 240
	heal = function(context, spec) -- 246
		for ____, target in ipairs(____exports.resolveTargets(context, spec.target)) do -- 247
			target.hp = target.hp + math.floor(spec.value + 0.5) -- 248
			if target.hp > target.maxHp then -- 248
				target.hp = target.maxHp -- 250
			end -- 250
		end -- 250
	end, -- 246
	dispel = function(context, spec) -- 254
		for ____, target in ipairs(____exports.resolveTargets(context, spec.target)) do -- 255
			target.debuffStacks = target.debuffStacks - math.floor(spec.value + 0.5) -- 256
			if target.debuffStacks < 0 then -- 256
				target.debuffStacks = 0 -- 258
			end -- 258
		end -- 258
	end, -- 254
	shield = function(context, spec) -- 262
		for ____, target in ipairs(____exports.resolveTargets(context, spec.target)) do -- 263
			target.shield = target.shield + math.floor(spec.value + 0.5) -- 264
		end -- 264
	end, -- 262
	buffDamage = function(context, spec) -- 267
		for ____, target in ipairs(____exports.resolveTargets(context, spec.target)) do -- 268
			target.buffStacks = target.buffStacks + math.floor(spec.value + 0.5) -- 269
		end -- 269
	end, -- 267
	debuffArmor = function(context, spec) -- 272
		for ____, target in ipairs(____exports.resolveTargets(context, spec.target)) do -- 273
			target.debuffStacks = target.debuffStacks + math.floor(spec.value + 0.5) -- 274
		end -- 274
	end, -- 272
	manaGain = function(context, spec) -- 277
		for ____, target in ipairs(____exports.resolveTargets(context, spec.target)) do -- 278
			____exports.addMana(target, spec.value) -- 279
		end -- 279
	end, -- 277
	boardShuffle = function(context, _spec) -- 282
		context.board:shuffle() -- 283
	end, -- 282
	boardBlast = function(context, _spec) -- 285
		context.board:blastLargestGroup() -- 286
	end, -- 285
	boardBlock = function(context, spec) -- 288
		context.board:blockCells(math.floor(spec.value + 0.5)) -- 289
	end -- 288
} -- 288
--- 尚未注册 handler 的效果次数（供自检与扩展性验收观测）。
local unhandledEffects = 0 -- 294
function ____exports.unhandledEffectCount() -- 296
	return unhandledEffects -- 297
end -- 296
function ____exports.resetUnhandledEffectCount() -- 300
	unhandledEffects = 0 -- 301
end -- 300
--- 统一执行器：只做“查表 → 按 target 取接收者 → 调用 handler”，
-- 返回实际执行的效果条目数；未注册的 kind 计入告警。
function ____exports.executeEffects(specs, context) -- 308
	local executed = 0 -- 309
	for ____, spec in ipairs(specs) do -- 310
		do -- 310
			local handler = effectHandlers[spec.kind] -- 311
			if handler == nil then -- 311
				unhandledEffects = unhandledEffects + 1 -- 313
				goto __continue64 -- 314
			end -- 314
			handler(context, spec) -- 316
			executed = executed + 1 -- 317
		end -- 317
		::__continue64:: -- 317
	end -- 317
	return executed -- 319
end -- 308
return ____exports -- 308